x-pack/legacy/plugins/integrations_manager/server/registry.ts

```diff
@@ -65,3 +72,85 @@ async function getOrFetchArchiveBuffer(key: string): Promise<Buffer> {
 async function fetchArchiveBuffer(key: string): Promise<Buffer> {
   return getResponseStream(`${REGISTRY}/package/${key}`).then(streamToBuffer);
 }
+
+export async function getObjects(
+  pkgkey: string,
+  filter = (entry: ArchiveEntry): boolean => true
+): Promise<SavedObject[]> {
+  // Create a Map b/c some values, especially index-patterns, are referenced multiple times
+  const objects: Map<string, SavedObject> = new Map();
+
+  // Get paths which match the given filter
+  const paths = await getArchiveInfo(`${pkgkey}.tar.gz`, filter);
+
+  // Add all objects which matched filter to the Map
+  paths.map(getObject).forEach(obj => objects.set(obj.id, obj));
+
+  // Each of those objects might have `references` property like [{id, type, name}]
+  for (const object of objects.values()) {
+    // For each of those objects
+    for (const reference of object.references) {
+      // Get the objects they reference. Call same function with a new filter
+      const referencedObjects = await getObjects(pkgkey, (entry: ArchiveEntry) => {
+        // Skip anything we've already stored
+        if (objects.has(reference.id)) return false;
+
+        // Is the archive entry the reference we want?
+        const { type, file } = pathParts(entry.path);
+        const isType = type === reference.type;
+        const isJson = file === `${reference.id}.json`;
+        return isType && isJson;
+      });
+
+      // Add referenced objects to the Map
+      referencedObjects.forEach(ro => objects.set(ro.id, ro));
+    }
+  }
+
+  // return the array of unique objects
+  return Array.from(objects.values());
+}
+
+export function pathParts(path: string) {
+  const [pkgkey, service, type, file] = path.split('/');
+
+  return { pkgkey, service, type, file };
+}
+
+// the assets from the registry are malformed
+// https://github.com/elastic/integrations-registry/issues/42
+function ensureJsonValues(obj: SavedObject) {
+  const { attributes } = obj;
+  if (
+    attributes.kibanaSavedObjectMeta &&
+    typeof attributes.kibanaSavedObjectMeta.searchSourceJSON !== 'string'
+  ) {
+    attributes.kibanaSavedObjectMeta.searchSourceJSON = JSON.stringify(
+      attributes.kibanaSavedObjectMeta.searchSourceJSON
+    );
+  }
+  ['optionsJSON', 'panelsJSON', 'uiStateJSON', 'visState']
+    .filter(key => typeof attributes[key] !== 'string')
+    .forEach(key => (attributes[key] = JSON.stringify(attributes[key])));
+  return obj;
+}
+
+function getAsset(key: string) {
+  const value = cacheGet(key);
+  if (value === undefined) throw new Error(`Cannot find asset ${key}`);
+
+  const json = value.toString('utf8');
+  const asset = JSON.parse(json);
+
+  return asset;
+}
+
+function getObject(key: string) {
+  const asset = getAsset(key);
+  const { type, file } = pathParts(key);
+
+  if (!asset.type) asset.type = type;
+  if (!asset.id) asset.id = file.replace('.json', '');
+
+  return ensureJsonValues(asset);
+}
```

Comment:
Mind adding a few more comments on these two functions to explain the diff between `getAsset` and `getObject`? I think I follow but a little lost on what's coming out of the cache that needs to be `toString`-ified.

Reply:
Updated in https://github.com/elastic/kibana/pull/40685/commits/5437febe0e26f5e7641e8895211346e27c81be9c